src/asmcnc/job/job_data.py

```diff
@@ -389,3 +400,209 @@ def post_job_data_update_post_send(self):
         self.post_production_notes = ''
         self.batch_number = ''
         self.percent_thru_job = 0
+
+
+    def read_from_recovery_file(self):
+        try:
+            with open(self.job_recovery_info_filepath, 'r') as job_recovery_info_file:
+                job_recovery_info = job_recovery_info_file.read().splitlines()
+
+            self.job_recovery_filepath = job_recovery_info[0]
+            self.job_recovery_cancel_line = int(job_recovery_info[1])
+
+        except:
+            print("Could not read recovery info")
+            print(str(traceback.format_exc()))
+
+    def write_to_recovery_file(self, cancel_line):
+        try:
+            # Account for number of lines added in by the software when running file
+            cancel_line -= self.job_recovery_offset
+            self.job_recovery_offset = 0
+
+            with open(self.job_recovery_info_filepath, 'w+') as job_recovery_info_file:
+                job_recovery_info_file.write(self.filename + "\n" + str(cancel_line))
+
+            # Simultaneously update variables
+            self.job_recovery_filepath = self.filename
+            self.job_recovery_cancel_line = cancel_line
+            self.job_recovery_selected_line = -1
+            self.job_recovery_gcode = []
+            self.job_recovery_offset = 0
+        
+        except:
+            print("Could not write recovery info")
+            print(str(traceback.format_exc()))
+
+    def clear_recovery_file(self):
+        try:
+            open(self.job_recovery_info_filepath, 'w').close()
+
+            # Simultaneously update variables
+            self.job_recovery_filepath = ''
+            self.job_recovery_cancel_line = 0
+            self.job_recovery_selected_line = -1
+            self.job_recovery_gcode = []
+            self.job_recovery_offset = 0
+        
+        except:
+            print("Could not clear recovery info")
+            print(str(traceback.format_exc()))
+
+
+    def reset_recovery(self):
+        self.job_recovery_selected_line = -1
+        self.job_recovery_gcode = []
+        self.job_recovery_offset = 0
+
+    def generate_recovery_gcode(self):
+        try:
+            recovery_gcode = []
+
+            # Recover modal gcodes
+
+            # Coordinate System Select
+            coord_system_line = next((s for s in reversed(self.job_gcode[:self.job_recovery_selected_line]) if re.search("G5[4-9]", s)), None)
+            if coord_system_line:
+                recovery_gcode.append(re.search("G5[4-9](\.[1-3])?", coord_system_line).group(0))
```

Comment:
If no G54, should it give user a warning? 

In this scenario, it would return True and a message

@Neex101

Reply:
maybe a warning message

Reply:
No warning required - have found various files in the gcode stash that don't explicitly state G54, so we can assume it's fine :)